### ISSUE TITLE
Provide a builder for EventGridPublisher

### ISSUE THREAD

**Issue by tomkerkhove** [arcus-azure/arcus.eventgrid#21]
**What feature would you like to have?**
We should provide a builder in order to create a `EventGridPublisher` which allows you to configure it.
This would allow us to provide flexibility towards consumers and allow us to provide extension points where it makes sense.

**Describe the solution you'd like**
A possible approach would be 

```csharp
var eventGridPublisher = EventGridPublisherBuilder
                              .ForTopic(topicEndpoint)
                              .WithAuthenticationKey(endpointKey) // Or provide capability to use Key Vault later on here
                              .Build()
```

Authentication is already one example of providing flexibility where they can choose what mechanism they want to use.

We can keep the current approach and mark it as obsolete which refers to the builder.

### Implementation specs
Mandatory specifications:
- Topic endpoint via `ForTopic`
- Authentication key via `UsingAuthenticationKey`

Example:
```csharp
var eventGridPublisher = EventGridPublisherBuilder
                               .ForTopic(topicEndpoint)
                               .UsingAuthenticationKey(<key>)
                               // or .UsingKeyVaultAuthentication(<url>, <app id>, <app key>)
                               .Create();
```

**Comment by SeanFeldman** [comment]
A minor suggestion:

```c#
var eventGridPublisher = EventGridPublisherBuilder
                              .ForTopic(topicEndpoint)
                              .WithAuthenticationKey(endpointKey) // Or provide capability to use Key Vault later on here
                              .Build()
var eventGridPublisher = EventGridPublisherBuilder
                              .ForTopic(topicEndpoint)
                              .UsingAuthenticationProvider(...)
                              .Build();
```

`.UsingAuthenticationProvider(...)` and have a default provider that is based on endpoint's key.

**Comment by murdockcrc** [comment]
The term Build is often associated with something else, I would suggest to use another term. It's only semantics, but I think we could be more precise instead with a "factory" or a "create"

**Comment by tomkerkhove** [comment]
That's a possibility but I would like to prefer to keep it as simple as possible.

However, just to pick your brain - How would you approach it then?
Something similar to this?
```csharp
var authenticationProvider = AuthenticationBuilder
                                                          .UsingApiKey(...)
                                                          .Build();

var eventGridPublisher = EventGridPublisherBuilder
                                                          .ForTopic(topicEndpoint)
                                                          .UsingAuthenticationProvider(authenticationProvider)
                                                          .Build();
```

**Comment by SeanFeldman** [comment]
I wouls construct an `IAuthenticate` and pass it in.
`IAuthenticate` variants could be `SASKeyAuthenticator`, `MSIAuthenticator`, etc

**Comment by stijnmoreels** [comment]
Just an idea: since the other issue #22 regarding the resiliency of the **Event Grid Publisher**, I seems to me that we should separate the "real" publishing from the more "wrapper"-like functionality (retryability, validation, security, ...) and not just writing everything in the `EventGridPublisher`.

This way, people could write their own "wrapper" for the publisher.

**Comment by tomkerkhove** [comment]
@stijnmoreels That would be a good suggestion indeed, but then I would do that as part of #22.
Does that make sense?

**Comment by stijnmoreels** [comment]
@tomkerkhove ah, yes; makes sense.

**Comment by stijnmoreels** [comment]
Am I correct that both the topic and authentication key are required? If so, it wouldn't make sense to me if we would allow to create a builder without calling these methods.
They should be provided in maybe a factory method at the start and from that other optional methods could be called (resilient, validation, custom, ...).

Or am I wrong on this?

**Comment by tomkerkhove** [comment]
That is correct - The builder should guide you through what is required and provide capability to specify optional information. This is something that they handle nicely with https://github.com/Azure/azure-libraries-for-net

Example - https://github.com/Azure-Samples/service-bus-dotnet-manage-queue-with-basic-features/blob/master/Program.cs

**Comment by tomkerkhove** [comment]
I've updated with the issue with specs for the first approach and is now ready for implementation.

**Comment by SeanFeldman** [comment]
> This is something that they handle nicely with https://github.com/Azure/azure-libraries-for-net

It's all nice and cute until one of those methods throws an exception.
NServiceBus used to heavily rely on the builder pattern to configure endpoints, but moved to more explicit option.

**Comment by tomkerkhove** [comment]
Thanks for the tip @SeanFeldman - In our scenario the only exceptions that will be thrown are argument exceptions if the input is not valid so I think we're ok given this will be a basic builder but will keep that tip in mind, thanks!

**Comment by tomkerkhove** [comment]
This is now available in v1.1.0 and documented in the [wiki](https://github.com/arcus-azure/arcus.eventgrid/wiki/Publishing-Events).
Thank you @stijnmoreels
